test/functional/test_framework/mininode.py

```diff
@@ -440,3 +454,127 @@ def network_thread_join(timeout=10):
+        hashList = [x.sha256 for x in headersList]
+
+        # Stop the list at the hash_stop header if found
+        index = len(headersList)
+        if (hash_stop in hashList):
+            index = hashList.index(hash_stop) + 1
+        response.headers = headersList[:index]
+
+        if response is not None:
+            self.send_message(response)
+
+    def on_reject(self, message):
+        """Store reject reason and code for testing."""
+        self.reject_code_received = message.code
+        self.reject_reason_received = message.reason
+
+    def send_blocks_and_test(self, blocks, rpc, success=True, reject_code=None, reject_reason=None, request_block=True):
+        """Send blocks to test node and test whether the tip advances.
+
+         - add all blocks to our block_store
+         - send a headers message for the final block
+         - the on_getheaders handler will ensure that any getheaders are responded to
+         - if request_block is True: wait for getdata for each of the blocks. The on_getdata handler will
+           ensure that any getdata messages are responded to
+         - if reject_code and reject_reason are set: assert that the correct reject message is received
+         - if success is True: assert that the node's tip advances to the most recent block.
+         - if success is False: assert that the node's tip doesn't advance."""
+
+        with mininode_lock:
+            self.reject_code_received = None
+            self.reject_reason_received = None
+
+        for block in blocks:
+            self.block_store[block.sha256] = block
+
+        self.send_message(msg_headers([blocks[-1]]))
+
+        if request_block:
+            wait_until(lambda: blocks[-1].sha256 in self.getdata_requests)
+
+        if success:
+            wait_until(lambda: rpc.getbestblockhash() == blocks[-1].hash)
+        else:
+            assert rpc.getbestblockhash() != blocks[-1].hash
+            if reject_code is not None:
+                wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock)
+            if reject_reason is not None:
+                wait_until(lambda: self.reject_reason_received == reject_reason, lock=mininode_lock)
+
+    def send_txs_and_test(self, txs, rpc, success=True, reject_code=None, reject_reason=None):
+        """Send txs to test node and test whether they're accepted to the mempool.
+
+         - add all txs to our tx_store
+         - send tx messages for all txs
+         - if reject_code and reject_reason are set: assert that the correct reject message is received
+         - if success is True: assert that the tx is accepted to the mempool
+         - if success is False: assert that the tx is not accepted to the mempool."""
+
+        for tx in txs:
+            self.tx_store[tx.sha256] = tx
+            self.send_message(msg_tx(tx))
+
+        self.sync_with_ping()
+
+        if success:
+            assert tx.hash in rpc.getrawmempool()
```

Comment:
The documentation states "whether *they* are accepted", but you only check one.

Reply:
fixed